### PR TITLE
[Maintenance] Bumps `re2` and `supertest`

### LINKED ISSUE #3283: CVE-2022-25901 (Medium) detected in cookiejar-2.1.3.tgz
## CVE-2022-25901 - Medium Severity Vulnerability
<details><summary> Vulnerable Library - <b>cookiejar-2.1.3.tgz</b></p></summary>

<p>simple persistent cookiejar system</p>
<p>Library home page: <a href="https://registry.npmjs.org/cookiejar/-/cookiejar-2.1.3.tgz">https://registry.npmjs.org/cookiejar/-/cookiejar-2.1.3.tgz</a></p>
<p>

Dependency Hierarchy:
  - supertest-6.2.2.tgz (Root Library)
    - superagent-7.1.2.tgz
      - :x: **cookiejar-2.1.3.tgz** (Vulnerable Library)
<p>Found in base branch: <b>main</b></p>
</p>
</details>
<p></p>
<details><summary> Vulnerability Details</summary>
<p>  
  
Versions of the package cookiejar before 2.1.4 are vulnerable to Regular Expression Denial of Service (ReDoS) via the Cookie.parse function, which uses an insecure regular expression.

<p>Publish Date: 2023-01-18
<p>URL: <a href=https://www.mend.io/vulnerability-database/CVE-2022-25901>CVE-2022-25901</a></p>
</p>
</details>
<p></p>
<details><summary> CVSS 3 Score Details (<b>5.3</b>)</summary>
<p>

Base Score Metrics:
- Exploitability Metrics:
  - Attack Vector: Network
  - Attack Complexity: Low
  - Privileges Required: None
  - User Interaction: None
  - Scope: Unchanged
- Impact Metrics:
  - Confidentiality Impact: None
  - Integrity Impact: None
  - Availability Impact: Low
</p>
For more information on CVSS3 Scores, click <a href="https://www.first.org/cvss/calculator/3.0">here</a>.
</p>
</details>
<p></p>
<details><summary> Suggested Fix</summary>
<p>

<p>Type: Upgrade version</p>
<p>Release Date: 2023-01-18</p>
<p>Fix Resolution: cookiejar - 2.1.4</p>

</p>
</details>
<p></p>

### PR DESCRIPTION
**Pull request by tmarkley**
### Description
* Removes manual resolution on `qs`. The latest version of `supertest` depends on v6.11.0.
* Adds steps for upgrading `re2`.
* Addresses CVE-2022-24999 (no issue opened).
 
### Issues Resolved

Resolves #3283 
Related to https://github.com/opensearch-project/OpenSearch-Dashboards/issues/1298
 
### Check List
- [x] All tests pass
  - [x] `yarn test:jest`
  - [x] `yarn test:jest_integration`
  - [x] `yarn test:ftr`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/dev/build/tasks/patch_native_modules_task.ts [hunk 1]
```diff
@@ -52,31 +52,39 @@ interface Package {
...
-        url: 'https://github.com/uhop/node-re2/releases/download/1.15.4/linux-x64-83.gz',
-        sha256: '1bbc3f90f0ba105772b37c04e3a718f69544b4df01dda00435c2b8e50b2ad0d9',
+        url: 'https://github.com/uhop/node-re2/releases/download/1.17.8/linux-x64-83.gz',
+        sha256: 'ac3810cefb4b3d732b01f8505082047bc943ed83ed32d6f955061b30351fefb4',
       },
       'linux-arm64': {
         url:
-          'https://d1v1sj258etie.cloudfront.net/node-re2/releases/download/1.15.4/linux-arm64-83.tar.gz',
-        sha256: 'f25124adc64d269a513b99abd4a5eed8d7a929db565207f8ece1f3b7b7931668',
+          'https://d1v1sj258etie.cloudfront.net/node-re2/releases/download/1.17.8/linux-arm64-83.tar.gz',
```
Comment (tmarkley) [inline src/dev/build/tasks/patch_native_modules_task.ts:80]: @AMoo-Miki is there a separate process for arm? Can we document that?

**Comment by tmarkley** [comment]
Dependency tree for `qs` after removing the manual resolution, this is how I identified the upstream dependencies that needed to be bumped:

```
$ yarn why qs
yarn why v1.22.15
[1/4] Why do we have the module "qs"...?
[2/4] Initialising dependency graph...
warning Resolution field "typescript@4.0.2" is incompatible with requested version "typescript@~4.5.2"
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "qs@6.11.0"
info Reasons this module exists
   - "_project_#supertest#superagent" depends on it
   - Hoisted from "_project_#supertest#superagent#qs"
info Disk size without dependencies: "284KB"
info Disk size with unique dependencies: "336KB"
info Disk size with transitive dependencies: "828KB"
info Number of shared dependencies: 7
=> Found "formidable#qs@6.9.3"
info This module exists because "_project_#supertest#superagent#formidable" depends on it.
info Disk size without dependencies: "236KB"
info Disk size with unique dependencies: "236KB"
info Disk size with transitive dependencies: "236KB"
info Number of shared dependencies: 0
=> Found "request#qs@6.5.3"
info This module exists because "_project_#re2#node-gyp#request" depends on it.
info Disk size without dependencies: "196KB"
info Disk size with unique dependencies: "196KB"
info Disk size with transitive dependencies: "196KB"
info Number of shared dependencies: 0
Done in 1.29s.

$ npm list qs
opensearch-dashboards@3.0.0 /workspaces/OpenSearch-Dashboards
├─┬ re2@1.15.4
│ └─┬ node-gyp@7.1.2
│   └─┬ request@2.88.2
│     └── qs@6.5.3 
└─┬ supertest@6.2.2
  └─┬ superagent@7.1.2
    ├─┬ formidable@2.0.1
    │ └── qs@6.9.3 
    └── qs@6.11.0 
```

File: src/dev/build/tasks/patch_native_modules_task.ts (hunk 1 above)
Comment (AMoo-Miki) [reply src/dev/build/tasks/patch_native_modules_task.ts:80]: There is a private document about the `re2` for `arm64`.

I also noticed that the `1.17.8` does not have a [Windows release](https://github.com/uhop/node-re2/releases) for Node 83. I will see if I can make one.

File: src/dev/build/tasks/patch_native_modules_task.ts [hunk 2]
```diff
@@ -52,31 +52,39 @@ interface Package {
...
+        sha256: '38ec2ac319ed7f5552d0b3c2785e9914acb40ee6ed6de2f605a22a6e6c0480ef',
         overriddenExtractMethod: 'untar',
         overriddenDestinationPath: 'node_modules/re2/build/Release',
       },
       'win32-x64': {
-        url: 'https://github.com/uhop/node-re2/releases/download/1.15.4/win32-x64-83.gz',
-        sha256: 'efe939d3cda1d64ee3ee3e60a20613b95166d55632e702c670763ea7e69fca06',
+        url: 'https://github.com/uhop/node-re2/releases/download/1.17.8/win32-x64-83.gz',
+        sha256: '16d6ab576b788495052c40d93854420fd5482719c0cf00bc7dde4c1ad632ef4b',
       },
```
Comment (AMoo-Miki) [inline src/dev/build/tasks/patch_native_modules_task.ts:88]: `win32-x64-83.gz` is not released for v1.17.8. We will either have to make and release our own or use a release that includes one.

**Review by AMoo-Miki — DISMISSED** [review]

File: src/dev/build/tasks/patch_native_modules_task.ts (hunk 2 above)
Comment (tmarkley) [reply src/dev/build/tasks/patch_native_modules_task.ts:88]: Thanks for calling that out Miki! I will look at the releases and see which one is the best option right now.

Comment (tmarkley) [reply src/dev/build/tasks/patch_native_modules_task.ts]: I've changed the version to `v1.17.4` which does include `win32-x64-83.gz` in the release.

File: src/dev/build/tasks/patch_native_modules_task.ts [hunk 3]
```diff
@@ -52,31 +52,42 @@ interface Package {
...
+// 1. Match `version` with the version in the yarn.lock file.
+// 2. Update the url to match the version.
+//    2a. If a Node.js update occurs, the node module version must match as
+//        well (i.e. '83'). See https://nodejs.org/en/download/releases/#ref-1.
+// 3. Generate the new checksum by executing the following commands:
+//    3a. `wget {url}`
+//    3b. `sha256sum {downloaded file name}`
+//    3c. For `linux-arm64`, the sha256 can also be found by replacing
+//        "linux-arm64-83.tar.gz" in the url with "sha256sum.txt.asc"
+//        and copying the sha256 from that file.
```
Comment (ashwin-pc) [inline src/dev/build/tasks/patch_native_modules_task.ts:65]: Thanks for adding these steps @tmarkley. Tangential question, why do we need this patching process?

Comment (tmarkley) [reply src/dev/build/tasks/patch_native_modules_task.ts:65]: Good question. This patch existed before the fork, but I know @ananzh did the work to bring the ARM build of re2 in. I haven't done a ton of digging in to why this exists, but I'd love to find an alternative approach.

**Comment by joshuarrrr** [comment]
> The versions in the yarn.lock are now v6.9.3 and v6.11.0 which do not have any current vulnerabilities.

@tmarkley - this is failing the whitesource check, because it says `6.9.3` does have vulnerabilities: https://www.mend.io/vulnerability-database/CVE-2022-24999

**Review by joshuarrrr — DISMISSED** [review]
We need a different `qs` version: https://www.mend.io/vulnerability-database/CVE-2022-24999

**Comment by ananzh** [comment]
@tmarkley 1.3.8 release is this week. Do you think we could still make it? or should we update the tag?

**Comment by tmarkley** [comment]
Taking a look now! Just getting back to things after taking an extended holiday break 😄

**Comment by tmarkley** [comment]
> > The versions in the yarn.lock are now v6.9.3 and v6.11.0 which do not have any current vulnerabilities.
> 
> @tmarkley - this is failing the whitesource check, because it says `6.9.3` does have vulnerabilities: https://www.mend.io/vulnerability-database/CVE-2022-24999

@joshuarrrr thanks for checking on that! Do we know why a security vulnerability issue wasn't automatically opened for that?

**Comment by AMoo-Miki** [comment]
>  Do we know why a security vulnerability issue wasn't automatically opened for that?

Mend was acting up for a few days; i think it is resolved now.

File: src/dev/build/tasks/patch_native_modules_task.ts (hunk 3 above)
Comment (AMoo-Miki) [reply src/dev/build/tasks/patch_native_modules_task.ts:65]: Installing re2 produces a binary for the specific platform architecture it was installed on. However, `yarn build-platform` can be used to create distributables for platforms other than the one the command is being executed on and for this, it downloads NodeJS executable and any binaries needed for the platforms.

**Review by AMoo-Miki — APPROVED** [review]

**Review by ananzh — APPROVED** [review]
LGTM gogogo!

**Comment by ananzh** [comment]
qs is now 6.11.0

```
qs@^6.11.0:
  version "6.11.0"
  resolved "https://registry.yarnpkg.com/qs/-/qs-6.11.0.tgz#fd0d963446f7a65e1367e01abd85429453f0c37a"
  integrity sha512-MvjoMCJwEarSbUYk5O+nmoSzSutSsTwF85zcHPQ9OrlFoZOYIjaqBAJIqIXjptyD5vThxGq52Xu/MaJzRkIk4Q==
```
The fix is https://github.com/ljharb/qs/pull/428/files
I checked their 6.11.0 tag and zip the code. Have verified manually that fix is there.

File: src/dev/build/tasks/patch_native_modules_task.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (tmarkley) [reply src/dev/build/tasks/patch_native_modules_task.ts:65]: Thanks for calling this out Miki! Here's the follow-up PR to clean this up :) https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3328

**Pull request merged**